src/Compilers/CSharp/Portable/Symbols/PublicModel/TypeSymbol.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel
+{
+    internal abstract class TypeSymbol : NamespaceOrTypeSymbol, ISymbol, ITypeSymbol
+    {
+        protected TypeSymbol(CodeAnalysis.NullableAnnotation nullableAnnotation)
+        {
+            NullableAnnotation = nullableAnnotation;
+        }
+
+        protected CodeAnalysis.NullableAnnotation NullableAnnotation { get; }
+
+        protected abstract ITypeSymbol WithNullableAnnotation(CodeAnalysis.NullableAnnotation nullableAnnotation);
+
+        internal abstract Symbols.TypeSymbol UnderlyingTypeSymbol { get; }
+
+        CodeAnalysis.NullableAnnotation ITypeSymbol.NullableAnnotation => NullableAnnotation;
+
+        ITypeSymbol ITypeSymbol.WithNullableAnnotation(CodeAnalysis.NullableAnnotation nullableAnnotation)
+        {
+            if (NullableAnnotation == nullableAnnotation)
+            {
+                return this;
+            }
+            else if (nullableAnnotation == UnderlyingTypeSymbol.DefaultNullableAnnotation)
+            {
+                return (TypeSymbol)UnderlyingSymbol.ISymbol;
+            }
+
+            return WithNullableAnnotation(nullableAnnotation);
+        }
+
+        bool ISymbol.IsDefinition
+        {
+            get
+            {
+                return (object)this == ((ISymbol)this).OriginalDefinition;
+            }
+        }
+
+        bool ISymbol.Equals(ISymbol other, CodeAnalysis.SymbolEqualityComparer equalityComparer)
+        {
+            return this.Equals(other as TypeSymbol, equalityComparer);
+        }
+
+        protected bool Equals(TypeSymbol otherType, CodeAnalysis.SymbolEqualityComparer equalityComparer)
+        {
+            if (otherType is null)
+            {
+                return false;
+            }
+            else if ((object)otherType == this)
+            {
+                return true;
+            }
+
+            var compareKind = equalityComparer.CompareKind;
+
+            if (NullableAnnotation != otherType.NullableAnnotation && (compareKind & TypeCompareKind.IgnoreNullableModifiersForReferenceTypes) == 0 &&
+                ((compareKind & TypeCompareKind.ObliviousNullableModifierMatchesAny) == 0 ||
+                 (NullableAnnotation != CodeAnalysis.NullableAnnotation.None && otherType.NullableAnnotation != CodeAnalysis.NullableAnnotation.None)) &&
+                 !(UnderlyingTypeSymbol.IsValueType && !UnderlyingTypeSymbol.IsNullableType()))
```

Comment:
Indentation is a little misleading here. This final comparison is on the top level, not part of the level it's currently indented to.